File: src/llm_old/tools/read_file.rs
use rig::completion::ToolDefinition;
use rig::tool::Tool;
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::path::{Path, PathBuf};
use std::sync::{Arc, RwLock};

/// Maximum number of lines that can be read from a file
const MAX_FILE_LINES: usize = 1000;

#[derive(Deserialize)]
pub struct ReadFileArgs {
  /// Path to the file to read (relative to cwd or absolute)
  path: String,
  /// Starting line number (0-indexed, optional)
  #[serde(default)]
  start_line: Option<usize>,
  /// Number of lines to read (optional, default: 100)
  #[serde(default)]
  max_lines: Option<usize>,
}

#[derive(Debug, thiserror::Error)]
pub enum ReadFileError {
  #[error("File not found: {0}")]
  FileNotFound(String),
  #[error("IO error: {0}")]
  IoError(#[from] std::io::Error),
  #[error("Invalid line range")]
  InvalidLineRange,
  #[error("Failed to acquire cwd lock")]
  LockError,
  #[error("Path traversal detected: {0}")]
  PathTraversal(String),
}

/// Tool for reading files by path
pub struct ReadFileTool {
  /// Current working directory
  cwd: Arc<RwLock<PathBuf>>,
}

impl ReadFileTool {
  pub fn new(cwd: Arc<RwLock<PathBuf>>) -> Self {
    Self { cwd }
  }

  /// Check if a path is safe (no path traversal outside cwd)
  fn is_safe_path(&self, path: &Path, cwd: &Path) -> bool {
    // Resolve the path and check if it's within cwd
    if let Ok(canonical) = path.canonicalize() {
      if let Ok(canonical_cwd) = cwd.canonicalize() {
        return canonical.starts_with(canonical_cwd);
      }
    }

    // If canonicalize fails (file doesn't exist yet), check components
    let resolved = if path.is_absolute() {
      path.to_path_buf()
    } else {
      cwd.join(path)
    };

    // Check for .. components that might escape cwd
    !resolved
      .components()
      .any(|c| matches!(c, std::path::Component::ParentDir))
  }
}

impl Tool for ReadFileTool {
  const NAME: &'static str = "read_file";

  type Args = ReadFileArgs;
  type Output = String;
  type Error = ReadFileError;

  async fn definition(&self, _prompt: String) -> ToolDefinition {
    ToolDefinition {
      name: "read_file".to_string(),
      description: "Read contents of a file by path. Supports reading specific line ranges to avoid overwhelming the context.".to_string(),
      parameters: json!({
        "type": "object",
        "properties": {
          "path": {
            "type": "string",
            "description": "Path to the file to read (relative to current working directory or absolute)"
          },
          "start_line": {
            "type": "integer",
            "description": "Starting line number (0-indexed, optional)",
            "minimum": 0
          },
          "max_lines": {
            "type": "integer",
            "description": format!("Maximum number of lines to read (default: 100, max: {})", MAX_FILE_LINES),
            "minimum": 1,
            "maximum": MAX_FILE_LINES
          }
        },
        "required": ["path"]
      }),
    }
  }

  async fn call(&self, args: Self::Args) -> Result<Self::Output, Self::Error> {
    let cwd = self.cwd.read().map_err(|_| ReadFileError::LockError)?;

    // Resolve the path
    let path = Path::new(&args.path);
    let full_path = if path.is_absolute() {
      path.to_path_buf()
    } else {
      cwd.join(path)
    };

    // Security check: prevent path traversal
    if !self.is_safe_path(&full_path, &cwd) {
      return Err(ReadFileError::PathTraversal(args.path));
    }

    // Check if file exists
    if !full_path.exists() {
      return Err(ReadFileError::FileNotFound(args.path));
    }

    // Read the file
    let content = std::fs::read_to_string(&full_path)?;
    let lines: Vec<&str> = content.lines().collect();

    // Apply line range
    let start_line = args.start_line.unwrap_or(0);
    let max_lines = args.max_lines.unwrap_or(100).min(MAX_FILE_LINES);

    if start_line >= lines.len() {
      return Err(ReadFileError::InvalidLineRange);
    }

    let end_line = (start_line + max_lines).min(lines.len());
    let selected_lines = &lines[start_line..end_line];

    let result = selected_lines.join("\n");
    let total_lines = lines.len();

    Ok(format!(
      "## File: {}\n\nShowing lines {}-{} of {} total lines:\n\n```\n{}\n```",
      args.path,
      start_line + 1,
      end_line,
      total_lines,
      result
    ))
  }
}

File: python/terminai_agent/tools/read_file.py
"""Tool for reading file contents."""

import logging
from pathlib import Path

from pydantic import BaseModel, Field

logger = logging.getLogger(__name__)

# Maximum number of lines that can be read from a file
MAX_FILE_LINES = 1000


class ReadFileArgs(BaseModel):
    """Arguments for read_file tool."""

    path: str = Field(
        description="Path to the file to read (relative to cwd or absolute)"
    )
    start_line: int | None = Field(
        None, description="Starting line number (0-indexed, optional)"
    )
    max_lines: int | None = Field(
        None,
        description=f"Maximum number of lines to read (default: 100, max: {MAX_FILE_LINES})",
    )


class ReadFileResult(BaseModel):
    """Result from read_file tool."""

    content: str = Field(description="File content or error message")
    path: str = Field(description="Path that was read")
    lines_shown: tuple[int, int] | None = Field(
        default=None, description="Range of lines shown (start, end) if partial read"
    )
    total_lines: int | None = Field(None, description="Total lines in file")
    error: str | None = Field(None, description="Error message if read failed")


def is_safe_path(path: Path, cwd: Path) -> bool:
    """Check if a path is safe (no path traversal outside cwd).

    Args:
        path: Path to check
        cwd: Current working directory

    Returns:
        True if path is safe, False otherwise
    """
    try:
        # Resolve the path and check if it's within cwd
        resolved = path.resolve()
        cwd_resolved = cwd.resolve()
        return resolved.is_relative_to(cwd_resolved)
    except (ValueError, OSError):
        return False


async def read_file(args: ReadFileArgs, cwd: str) -> ReadFileResult:
    """Read contents of a file by path.

    Args:
        args: Tool arguments
        cwd: Current working directory

    Returns:
        ReadFileResult with file content or error
    """
    cwd_path = Path(cwd)
    file_path = Path(args.path)

    # Resolve the path
    if file_path.is_absolute():
        full_path = file_path
    else:
        full_path = cwd_path / file_path

    # Security check: prevent path traversal
    if not is_safe_path(full_path, cwd_path):
        return ReadFileResult(
            content="",
            path=args.path,
            error=f"Path traversal detected: {args.path}",
        )

    # Check if file exists
    if not full_path.exists():
        return ReadFileResult(
            content="",
            path=args.path,
            error=f"File not found: {args.path}",
        )

    # Check if it's a file (not directory)
    if not full_path.is_file():
        return ReadFileResult(
            content="",
            path=args.path,
            error=f"Not a file: {args.path}",
        )

    try:
        # Read the file
        content = full_path.read_text()
        lines = content.splitlines()

        # Apply line range
        start_line = args.start_line or 0
        max_lines = min(args.max_lines or 100, MAX_FILE_LINES)

        if start_line >= len(lines):
            return ReadFileResult(
                content="",
                path=args.path,
                error=f"Invalid start_line: {start_line} (file has {len(lines)} lines)",
            )

        end_line = min(start_line + max_lines, len(lines))
        selected_lines = lines[start_line:end_line]

        result_content = "\n".join(selected_lines)
        total_lines = len(lines)

        # Format output
        if start_line > 0 or end_line < total_lines:
            # Partial read
            output = f"## File: {args.path}\n\n"
            output += f"Showing lines {start_line + 1}-{end_line} of {total_lines} total lines:\n\n"
            output += f"```\n{result_content}\n```"
        else:
            # Full file
            output = f"## File: {args.path}\n\n```\n{result_content}\n```"

        return ReadFileResult(
            content=output,
            path=args.path,
            lines_shown=(start_line + 1, end_line),
            total_lines=total_lines,
        )

    except UnicodeDecodeError:
        return ReadFileResult(
            content="",
            path=args.path,
            error=f"Cannot read binary file: {args.path}",
        )
    except PermissionError:
        return ReadFileResult(
            content="",
            path=args.path,
            error=f"Permission denied: {args.path}",
        )
    except Exception as e:
        logger.exception(f"Error reading file {args.path}")
        return ReadFileResult(
            content="",
            path=args.path,
            error=f"Error reading file: {e}",
        )
